### PR TITLE
Bugfix for cas insensitive searches for a file path

### LINKED ISSUE #1587: Case insensitive search for a file does not work
Imagine you have a file at `/path/to/library/A/File.mp3`. Executing `beet ls -f $path` will give us
```
/path/to/library/A/File.mp3
```
So I would expect to get a result when executing `beet ls path:"/path/to/library/A/File.mp3"`, but beets doesn't find the file.

I tried to find the bug but didn't succeed. One thing I noted is the where clause of the `PathQuery` (line 103 of `library.py`): The bound parameter is in lower case but the data at the database might contain upper case letters. Therefore, a check like `path = ?` might fail even if it shouldn't. So I changed it to `lower(path) = ?` but that didn't change the described behavior.

What am I doing wrong?

### PR DESCRIPTION
**Pull request by mried**
This will fix #1587 by using the case-lowering of the database instead of Python and SQLite.

### CODE REVIEW

File: docs/changelog.rst [hunk 1]
```diff
@@ -43,6 +43,8 @@ Fixes:
   written to files. Thanks to :user:`jdetrey`. :bug:`1303` :bug:`1589`
 * :doc:`/plugins/replaygain`: Avoid a crash when the PyAudioTools backend
   encounters an error. :bug:`1592`
+* Case insensitive searches might have returned nothing because of a wrong
+  SQL query.
```
Comment (sampsyo) [inline docs/changelog.rst]: For specificity: "Case insensitive searches" -> "Case-insensitive path queries"

File: beets/library.py [hunk 2]
```diff
@@ -66,10 +66,6 @@ def __init__(self, field, pattern, fast=True, case_sensitive=None):
             case_sensitive = platform.system() != 'Windows'
         self.case_sensitive = case_sensitive
 
-        # Use a normalized-case pattern for case-insensitive matches.
-        if not case_sensitive:
-            pattern = pattern.lower()
-
```
Comment (sampsyo) [inline beets/library.py]: It occurs to me that, if we remove this, we'll also need to change the `match` method below (which currently assumes both `self.file_path` and `self.dir_path` are lower-cased. Maybe we should just leave this in.

**Comment by sampsyo** [comment]
I think I finally see what the problem is here: `lower()` in SQLite transparently converts BLOBs to strings. See:

    sqlite> select quote(x'034445');
    X'034445'
    sqlite> select quote(lower(x'034445'));
    'de'

Meaning that calling `lower()` on the data in the table is producing a string, but we're still passing in a BLOB containing the lower-cased query string from Python. SQLite considers a string/BLOB comparison to be false.

    >>> c.execute('select ? = ?', (buffer("hi"), "hi")).fetchone()[0]
    0
    >>> c.execute('select lower(?) = ?', (buffer("hi"), "hi")).fetchone()[0]
    1

It would be *really nice*, then, if we could just pass in the `file_path` string as a string, not a blob, and do the comparison without any more conversions. But SQLite won't let us convert an arbitrary byte sequence to a string. :cry:

In fact, we're already inviting trouble by calling `lower(?)` on the BLOB data stored in the database: if it doesn't match SQLite's internal encoding (which it certainly might not! We don't even know which encoding SQLite was compiled with), converting to a string causes an error:

    >>> c.execute('select lower(?)', (buffer(u'caf\xe9'.encode('latin1')),)).fetchone()

(That's on my SQLite, which happens to have a UTF-8 internal encoding. This Latin1-encoded string happens to be invalid UTF-8.)

Anyway, I think the conclusion to all of this is that we can't rely on SQLite's internal BLOB-to-TEXT, which is implicitly invoked by `lower()`. I think the only alternative is to sort out how to use `LIKE` to get its case-insensitive behavior.

**Comment by sampsyo** [comment]
Note that `LIKE` seems to work fine with arbitrary binary BLOBs:

    >>> c.execute('select ? like ?', (buffer(u'caf\xe9'.encode('latin1')), buffer('caf%'))).fetchone()[0]
    1

And that it is indeed case-insensitive, even for BLOBs:

    >>> c.execute('select ? like ?', (buffer('X'), buffer('x'))).fetchone()[0]
    1

Woohoo!

File: beets/library.py (hunk 2 above)
Comment (mried) [inline beets/library.py]: Ehm... yes, you're absolutely right. I must have overlooked this usage :open_mouth:...

**Comment by mried** [comment]
Everything you mentioned is now implemented and seems to work perfectly :smile:

File: beets/library.py [hunk 3]
```diff
@@ -102,8 +101,11 @@ def col_clause(self):
         escape = lambda m: self.escape_char + m.group(0)
         dir_pattern = self.escape_re.sub(escape, self.dir_path)
         dir_blob = buffer(dir_pattern + b'%')
-        return '({0} = ?) || ({0} LIKE ? ESCAPE ?)'.format(self.field), \
-               (file_blob, dir_blob, self.escape_char)
+        file_pattern = self.escape_re.sub(escape, self.file_path)
+        file_blob = buffer(file_pattern + b'%')
```
Comment (sampsyo) [inline beets/library.py]: I think we don't want the `+ b'%'` here, which will let the query `path:/my/song.mp3` match a file at path `/my/song.mp3.foo`. The `+ b'%'` is only needed for the directory part of the query (`dir_pattern`).

**Comment by sampsyo** [comment]
AWESOME! :sparkles: I'm so glad this works. This is a nice, simple solution.

One tiny comment above; then feel free to hit the merge button if you like. :rocket:

**Pull request merged**